File: apina-core/src/main/java/fi/evident/apina/spring/JacksonTypeTranslator.java
package fi.evident.apina.spring;

import fi.evident.apina.java.model.*;
import fi.evident.apina.java.model.type.*;
import fi.evident.apina.model.ApiDefinition;
import fi.evident.apina.model.ClassDefinition;
import fi.evident.apina.model.EnumDefinition;
import fi.evident.apina.model.PropertyDefinition;
import fi.evident.apina.model.settings.TranslationSettings;
import fi.evident.apina.model.type.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.function.Predicate;

import static fi.evident.apina.spring.NameTranslator.translateClassName;
import static fi.evident.apina.utils.PropertyUtils.propertyNameForGetter;
import static java.util.Arrays.asList;
import static java.util.Objects.requireNonNull;

/**
 * Translates Java types to model types.
 */
final class JacksonTypeTranslator {

    private final TranslationSettings settings;
    private final ClassMetadataCollection classes;
    private final ApiDefinition api;

    /**
     * Maps translated simple names back to their original types.
     * Needed to make sure that our mapping remains unique.
     */
    private final Map<String,JavaBasicType> translatedNames = new HashMap<>();

    private static final Logger log = LoggerFactory.getLogger(JacksonTypeTranslator.class);

    private static final JavaBasicType JSON_IGNORE = new JavaBasicType("com.fasterxml.jackson.annotation.JsonIgnore");
    private static final JavaBasicType JSON_VALUE = new JavaBasicType("com.fasterxml.jackson.annotation.JsonValue");
    private static final List<JavaBasicType> OPTIONAL_NUMBER_TYPES =
            asList(new JavaBasicType(OptionalInt.class), new JavaBasicType(OptionalLong.class), new JavaBasicType(OptionalDouble.class));

    public JacksonTypeTranslator(TranslationSettings settings, ClassMetadataCollection classes, ApiDefinition api) {
        this.settings = requireNonNull(settings);
        this.classes = requireNonNull(classes);
        this.api = requireNonNull(api);
    }

    public ApiType translateType(JavaType javaType, TypeEnvironment env) {
        return javaType.accept(new JavaTypeVisitor<TypeEnvironment, ApiType>() {
            @Override
            public ApiType visit(JavaArrayType type, TypeEnvironment ctx) {
                return new ApiArrayType(translateType(type.getElementType(), ctx));
            }

            @Override
            public ApiType visit(JavaBasicType type, TypeEnvironment ctx) {
                if (classes.isInstanceOf(type, Collection.class)) {
                    return new ApiArrayType(ApiPrimitiveType.ANY);

                } else if (classes.isInstanceOf(type, Map.class)) {
                    return new ApiDictionaryType(ApiPrimitiveType.ANY);

                } else if (type.equals(new JavaBasicType(String.class))) {
                    return ApiPrimitiveType.STRING;

                } else if (classes.isNumber(type)) {
                    return ApiPrimitiveType.NUMBER;

                } else if (type.equals(JavaBasicType.BOOLEAN) || type.equals(new JavaBasicType(Boolean.class))) {
                    return ApiPrimitiveType.BOOLEAN;

                } else if (OPTIONAL_NUMBER_TYPES.contains(type)) {
                    return ApiPrimitiveType.NUMBER;

                } else if (type.equals(new JavaBasicType(Object.class))) {
                    return ApiPrimitiveType.ANY;

                } else if (type.isVoid()) {
                    return ApiPrimitiveType.VOID;

                } else {
                    return translateClassType(type, ctx);
                }
            }

            @Override
            public ApiType visit(JavaParameterizedType type, TypeEnvironment ctx) {
                JavaType baseType = type.getBaseType();
                List<JavaType> arguments = type.getArguments();

                if (classes.isInstanceOf(baseType, Collection.class) && arguments.size() == 1)
                    return new ApiArrayType(translateType(arguments.get(0), ctx));
                else if (classes.isInstanceOf(baseType, Map.class) && arguments.size() == 2 && classes.isInstanceOf(arguments.get(0), String.class))
                    return new ApiDictionaryType(translateType(arguments.get(1), ctx));
                else if (classes.isInstanceOf(baseType, Optional.class) && arguments.size() == 1)
                    return translateType(arguments.get(0), ctx);
                else
                    return translateType(baseType, ctx); // TODO: use arguments
            }

            @Override
            public ApiType visit(JavaTypeVariable type, TypeEnvironment ctx) {
                return ctx.lookup(type).map(t -> translateType(t, ctx)).orElse(ApiPrimitiveType.ANY);
            }

            @Override
            public ApiType visit(JavaWildcardType type, TypeEnvironment ctx) {
                return type.getLowerBound().map(t -> translateType(t, ctx)).orElse(ApiPrimitiveType.ANY);
            }

            @Override
            public ApiType visit(JavaInnerClassType type, TypeEnvironment ctx) {
                throw new UnsupportedOperationException("translating inner class types is not supported: " + type);
            }
        }, env);
    }

    private ApiType translateClassType(JavaBasicType type, TypeEnvironment env) {
        ApiTypeName typeName = classNameForType(type);

        if (settings.isImported(typeName))
            return new ApiBlackBoxType(typeName);

        if (settings.isBlackBoxClass(type.getName()) || hasJsonValueAnnotation(type)) {
            log.debug("Translating {} as black box", type.getName());

            ApiBlackBoxType blackBoxType = new ApiBlackBoxType(typeName);
            api.addBlackBox(typeName);
            return blackBoxType;
        }

        ApiClassType classType = new ApiClassType(typeName);

        if (!api.containsType(typeName)) {
            JavaClass aClass = classes.findClass(type.getName()).orElse(null);
            if (aClass != null) {
                if (aClass.isEnum()) {
                    api.addEnumDefinition(new EnumDefinition(typeName, aClass.getEnumConstants()));

                } else {
                    ClassDefinition classDefinition = new ClassDefinition(typeName);

                    // We must first add the definition to api and only then proceed to
                    // initialize it because initialization of properties could refer
                    // back to this same class and we'd get infinite recursion if the
                    // class is not already installed.
                    api.addClassDefinition(classDefinition);
                    initClassDefinition(classDefinition, new BoundClass(aClass, env));
                }
            }
        }

        return classType;
    }

    private ApiTypeName classNameForType(JavaBasicType type) {
        String translatedName = translateClassName(type.getName());

        JavaBasicType existingType = translatedNames.putIfAbsent(translatedName, type);
        if (existingType != null && !type.equals(existingType))
            throw new DuplicateClassNameException(type.getName(), existingType.getName());

        return new ApiTypeName(translatedName);
    }

    private boolean hasJsonValueAnnotation(JavaBasicType type) {
        return classes.findClass(type.getName()).map(cl -> cl.hasMethodWithAnnotation(JSON_VALUE)).orElse(false);
    }

    private void initClassDefinition(ClassDefinition classDefinition, BoundClass boundClass) {
        Set<String> ignoredProperties = ignoredProperties(boundClass);

        Predicate<String> acceptProperty = name -> !classDefinition.hasProperty(name) && !ignoredProperties.contains(name);

        for (BoundClass cl : classesUpwardsFrom(boundClass)) {
            addPropertiesFromGetters(cl, classDefinition, acceptProperty);
            addPropertiesFromFields(cl, classDefinition, acceptProperty);
        }
    }

    private void addPropertiesFromFields(BoundClass boundClass, ClassDefinition classDefinition, Predicate<String> acceptProperty) {
        for (JavaField field : boundClass.getJavaClass().getPublicInstanceFields()) {
            String name = field.getName();
            if (acceptProperty.test(name)) {
                ApiType type = translateType(field.getType(), boundClass.getEnvironment());
                classDefinition.addProperty(new PropertyDefinition(name, type));
            }
        }
    }

    private void addPropertiesFromGetters(BoundClass boundClass, ClassDefinition classDefinition, Predicate<String> acceptProperty) {
        for (JavaMethod getter : boundClass.getJavaClass().getGetters()) {
            String name = propertyNameForGetter(getter.getName());
            if (acceptProperty.test(name)) {
                ApiType type = translateType(getter.getReturnType(), boundClass.getEnvironment());
                classDefinition.addProperty(new PropertyDefinition(name, type));
            }
        }
    }

    private Set<String> ignoredProperties(BoundClass type) {
        Set<String> ignores = new HashSet<>();

        List<BoundClass> classes = classesUpwardsFrom(type);

        for (int i = classes.size() - 1; i >= 0; i--) {
            JavaClass aClass = classes.get(i).getJavaClass();

            for (JavaField field : aClass.getPublicInstanceFields()) {
                if (field.findAnnotation(JSON_IGNORE).map(JacksonTypeTranslator::isIgnore).orElse(false))
                    ignores.add(field.getName());
                else
                    ignores.remove(field.getName());
            }

            for (JavaMethod getter : aClass.getGetters()) {
                JavaAnnotation ignore = getter.findAnnotation(JSON_IGNORE).orElse(null);
                if (ignore != null) {
                    String name = propertyNameForGetter(getter.getName());
                    if (isIgnore(ignore)) {
                        ignores.add(name);
                    } else {
                        ignores.remove(name);
                    }
                }
            }
        }

        return ignores;
    }

    private static boolean isIgnore(JavaAnnotation ignore) {
        return ignore.getAttribute("value", Boolean.class).orElse(true);
    }

    private List<BoundClass> classesUpwardsFrom(BoundClass javaClass) {
        List<BoundClass> result = new ArrayList<>();

        for (BoundClass cl = javaClass; cl != null; cl = boundClassFor(cl.getJavaClass().getSuperClass(), cl.getEnvironment()).orElse(null))
            addClassAndInterfacesAt(cl, result);

        return result;
    }

    private void addClassAndInterfacesAt(BoundClass c, List<BoundClass> result) {
        if (result.stream().anyMatch(bc -> c.getJavaClass().equals(bc.getJavaClass())))
            return;

        result.add(c);

        for (JavaType interfaceType : c.getJavaClass().getInterfaces()) {
            BoundClass interfaceClass = boundClassFor(interfaceType, c.getEnvironment()).orElse(null);
            if (interfaceClass != null)
                addClassAndInterfacesAt(interfaceClass, result);
        }
    }

    private Optional<BoundClass> boundClassFor(JavaType type, TypeEnvironment env) {
        return classes.findClass(type.getNonGenericClassName()).map(c -> {
            if (type instanceof JavaParameterizedType) {
                JavaParameterizedType parameterizedType = (JavaParameterizedType) type;
                List<JavaType> args = env.resolve(parameterizedType.getArguments());
                return new BoundClass(c, c.getSchema().apply(args));
            } else {
                return new BoundClass(c, TypeEnvironment.empty());
            }
        });
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/spring/JacksonTypeTranslator.kt
package fi.evident.apina.spring

import fi.evident.apina.java.model.*
import fi.evident.apina.java.model.type.BoundClass
import fi.evident.apina.java.model.type.JavaType
import fi.evident.apina.java.model.type.TypeEnvironment
import fi.evident.apina.model.*
import fi.evident.apina.model.type.ApiType
import fi.evident.apina.model.type.ApiTypeName

/**
 * Translates Jackson classes to model types.
 */
internal class JacksonTypeTranslator(
    private val typeTranslator: TypeTranslator,
    private val classes: JavaModel,
    private val api: ApiDefinition
) {

    fun translateClass(javaClass: JavaClass, typeName: ApiTypeName, env: TypeEnvironment): ApiType {
        val jsonValueMethod = javaClass.findMethodWithAnnotation(JSON_VALUE)
        if (jsonValueMethod != null) {
            api.addTypeAlias(typeName, typeTranslator.translateType(jsonValueMethod.returnType, env))
            return ApiType.BlackBox(typeName)
        }

        val classType = ApiType.Class(typeName)

        if (!api.containsType(typeName)) {
            when {
                javaClass.isEnum ->
                    api.addEnumDefinition(EnumDefinition(typeName, javaClass.enumConstants))
                javaClass.hasAnnotation(JSON_TYPE_INFO) -> {
                    val typeInfo =
                        javaClass.findAnnotation(JSON_TYPE_INFO) ?: error("@JsonTypeInfo missing for $javaClass")
                    createDiscriminatedUnion(javaClass, typeInfo)
                }
                else -> {
                    val classDefinition = ClassDefinition(typeName)

                    // We must first add the definition to api and only then proceed to
                    // initialize it because initialization of properties could refer
                    // back to this same class and we'd get infinite recursion if the
                    // class is not already installed.
                    api.addClassDefinition(classDefinition)
                    initClassDefinition(classDefinition, BoundClass(javaClass, env))
                }
            }
        }

        return classType
    }

    private fun createDiscriminatedUnion(javaClass: JavaClass, typeInfo: JavaAnnotation) {
        val use = typeInfo.getRequiredAttribute<EnumValue>("use").constant
        val include = typeInfo.getAttribute<EnumValue>("include")?.constant ?: "PROPERTY"
        val property = typeInfo.getAttribute("property") ?: ""

        check(use == "NAME") { "Only 'use=NAME' is supported for @JsonTypeInfo (in ${javaClass.name})" }
        check(include == "PROPERTY") { "Only 'include=PROPERTY' is supported for @JsonTypeInfo (in ${javaClass.name})" }
        check(property.isNotEmpty()) { "No 'property' defined for @JsonTypeInfo (in ${javaClass.name})" }

        val union = DiscriminatedUnionDefinition(typeTranslator.classNameForType(javaClass.type.toBasicType()), property)

        // Add this before further processing because the type may be recursively referenced inside subclasses
        api.addDiscriminatedUnion(union)

        for ((name, cl) in findSubtypes(javaClass)) {
            val def = ClassDefinition(typeTranslator.classNameForType(cl.type.toBasicType()))
            initClassDefinition(def, BoundClass(cl, TypeEnvironment.empty()))
            union.addType(name, def)
        }
    }

    private fun findSubtypes(javaClass: JavaClass): List<Pair<String, JavaClass>> {
        val subTypes = javaClass.findAnnotation(JSON_SUB_TYPES)
        if (subTypes != null) {
            return subTypes.getAttributeValues("value").map { subType ->
                val annotation = subType as JavaAnnotation
                val value = annotation.getRequiredAttribute<JavaType>("value")
                val name = annotation.getAttribute<String>("name")
                    ?: error("no name defined subclass $value (in ${javaClass.type})")
                val cl = classes.findClass(value.toBasicType()) ?: error("could not find class $value")
                name to cl
            }
        } else {
            // If there's no @JsonSubTypes annotation, we can still do the same thing as Jackson does and detect
            // subtypes of sealed class automatically. We could scan Kotlin's metadata, but there's actually a simpler
            // way: we know that sealed classes must reside in the same package as the original class since they must
            // be defined in the same source file and one source file can only output classes in one package. Therefore
            // we can just scan the classes in the same package as the original class.
            return classes.findDirectSubclassesInSamePackage(javaClass).map { cl ->
                val typeName = cl.findAnnotation(JSON_TYPE_NAME)
                    ?: error("No @JsonTypeName annotation found for ${cl.name}. Either specify @JsonTypeName or use @JsonSubTypes at ${javaClass.name}")
                val name = typeName.getRequiredAttribute<String>("value")
                name to cl
            }
        }
    }

    private fun initClassDefinition(
        classDefinition: ClassDefinition,
        boundClass: BoundClass,
        prefix: String = "",
        suffix: String = ""
    ) {
        val ignoredProperties = ignoredProperties(boundClass)

        val acceptProperty = { name: String -> !classDefinition.hasProperty(name) && name !in ignoredProperties }

        for (cl in classes.classesUpwardsFrom(boundClass)) {
            if (cl.javaClass.isRecord) {
                for (component in cl.javaClass.recordComponents)
                    processProperty(cl, classDefinition, component.name, component, component.type, acceptProperty, prefix, suffix)
            } else {
                for (getter in cl.javaClass.getters)
                    processProperty(cl, classDefinition, getter.propertyName, getter, getter.returnType, acceptProperty, prefix, suffix)
                for (field in cl.javaClass.publicInstanceFields)
                    processProperty(cl, classDefinition, field.name, field, field.type, acceptProperty, prefix, suffix)
            }
        }
    }

    private fun processProperty(
        boundClass: BoundClass,
        classDefinition: ClassDefinition,
        name: String,
        element: JavaAnnotatedElement,
        javaType: JavaType,
        acceptProperty: (String) -> Boolean,
        prefix: String,
        suffix: String
    ) {
        val unwrappedAnnotation = element.findAnnotation(JSON_UNWRAPPED)
        if (unwrappedAnnotation != null && unwrappedAnnotation.getAttribute("enabled", Boolean::class.java) != false) {
            val newPrefix = unwrappedAnnotation.getAttribute("prefix", String::class.java) ?: ""
            val newSuffix = unwrappedAnnotation.getAttribute("suffix", String::class.java) ?: ""
            val type = classes.findClass(javaType.toBasicType())
            if (type != null)
                initClassDefinition(
                    classDefinition,
                    BoundClass(type, TypeEnvironment.empty()),
                    prefix = newPrefix + prefix,
                    suffix = suffix + newSuffix
                )

        } else if (acceptProperty(name)) {
            val type = typeTranslator.translateType(javaType, element, boundClass.environment)
            classDefinition.addProperty(PropertyDefinition(prefix + name + suffix, type))
        }
    }

    private fun ignoredProperties(type: BoundClass): Set<String> {
        val ignores = HashSet<String>()

        val classes = classes.classesUpwardsFrom(type)

        for (i in classes.indices.reversed()) {
            val aClass = classes[i].javaClass

            for (field in aClass.publicInstanceFields) {
                if (field.findAnnotation(JSON_IGNORE)
                        ?.isIgnore() == true || field.isTransient || field.hasExternalIgnoreAnnotation()
                )
                    ignores.add(field.name)
                else
                    ignores.remove(field.name)
            }

            for (getter in aClass.getters) {
                val ignore =
                    getter.findAnnotation(JSON_IGNORE) ?: getter.correspondingField?.findAnnotation(JSON_IGNORE)

                if (ignore != null) {
                    if (ignore.isIgnore()) {
                        ignores.add(getter.propertyName)
                    } else {
                        ignores.remove(getter.propertyName)
                    }
                } else if (getter.hasExternalIgnoreAnnotation() || getter.correspondingField?.hasExternalIgnoreAnnotation() == true) {
                    ignores.add(getter.propertyName)
                }
            }
        }

        return ignores
    }

    private fun JavaAnnotatedElement.hasExternalIgnoreAnnotation() =
        hasAnnotation(JAVA_BEANS_TRANSIENT) || hasAnnotation(SPRING_DATA_TRANSIENT)

    companion object {

        private val JSON_IGNORE = JavaType.Basic("com.fasterxml.jackson.annotation.JsonIgnore")
        private val JSON_VALUE = JavaType.Basic("com.fasterxml.jackson.annotation.JsonValue")
        private val JSON_TYPE_INFO = JavaType.Basic("com.fasterxml.jackson.annotation.JsonTypeInfo")
        private val JSON_TYPE_NAME = JavaType.Basic("com.fasterxml.jackson.annotation.JsonTypeName")
        private val JSON_SUB_TYPES = JavaType.Basic("com.fasterxml.jackson.annotation.JsonSubTypes")
        private val JSON_UNWRAPPED = JavaType.Basic("com.fasterxml.jackson.annotation.JsonUnwrapped")
        private val JAVA_BEANS_TRANSIENT = JavaType.Basic(java.beans.Transient::class)
        private val SPRING_DATA_TRANSIENT = JavaType.Basic("org.springframework.data.annotation.Transient")

        private fun JavaAnnotation.isIgnore() = getAttribute("value") ?: true
    }
}
